drivers/soundwire/intel.c

```diff
@@ -1081,44 +1071,81 @@ static int intel_init(struct sdw_intel *sdw)
 }
 
 /*
- * probe and init
+ * probe, startup and remove callbacks
  */
-static int intel_probe(struct platform_device *pdev)
+static int intel_master_probe(struct sdw_intel_master_dev *master_dev)
 {
-	struct sdw_cdns_stream_config config;
+	struct device *dev = sdw_intel_master_to_dev(master_dev);
 	struct sdw_intel *sdw;
+	struct sdw_bus *bus;
 	int ret;
 
-	sdw = devm_kzalloc(&pdev->dev, sizeof(*sdw), GFP_KERNEL);
+	sdw = devm_kzalloc(dev, sizeof(*sdw), GFP_KERNEL);
 	if (!sdw)
 		return -ENOMEM;
 
-	sdw->instance = pdev->id;
-	sdw->link_res = dev_get_platdata(&pdev->dev);
-	sdw->cdns.dev = &pdev->dev;
+	master_dev->pdata = sdw;
+
+	sdw->instance = master_dev->link_id;
+	sdw->link_res = master_dev->link_res;
+	sdw->cdns.dev = dev;
 	sdw->cdns.registers = sdw->link_res->registers;
-	sdw->cdns.instance = sdw->instance;
+	sdw->cdns.instance = master_dev->link_id;
 	sdw->cdns.msg_count = 0;
-	sdw->cdns.bus.dev = &pdev->dev;
-	sdw->cdns.bus.link_id = pdev->id;
+	sdw->cdns.bus.link_id = master_dev->link_id;
 
 	sdw_cdns_probe(&sdw->cdns);
 
 	/* Set property read ops */
+	bus = &sdw->cdns.bus;
 	sdw_intel_ops.read_prop = intel_prop_read;
-	sdw->cdns.bus.ops = &sdw_intel_ops;
+	bus->ops = &sdw_intel_ops;
 
-	platform_set_drvdata(pdev, sdw);
+	/* set driver data, accessed by snd_soc_dai_set_drvdata() */
+	dev_set_drvdata(dev, &sdw->cdns);
 
-	ret = sdw_add_bus_master(&sdw->cdns.bus);
+	ret = sdw_bus_master_add(bus, dev, master_dev->fwnode);
 	if (ret) {
-		dev_err(&pdev->dev, "sdw_add_bus_master fail: %d\n", ret);
+		dev_err(dev, "sdw_add_bus_master fail: %d\n", ret);
 		return ret;
 	}
 
-	if (sdw->cdns.bus.prop.hw_disabled) {
-		dev_info(&pdev->dev, "SoundWire master %d is disabled, ignoring\n",
-			 sdw->cdns.bus.link_id);
+	if (bus->prop.hw_disabled)
+		dev_info(dev,
+			 "SoundWire master %d is disabled, will be ignored\n",
+			 bus->link_id);
+
+	/* Acquire IRQ */
+	ret = request_threaded_irq(sdw->link_res->irq,
+				   sdw_cdns_irq, sdw_cdns_thread,
+				   IRQF_SHARED, KBUILD_MODNAME, &sdw->cdns);
+	if (ret < 0) {
+		dev_err(dev, "unable to request IRQ %d, disabling device\n",
+			sdw->link_res->irq);
+		goto err_init;
+	}
+
+	complete(&master_dev->probe_complete);
+
+	return 0;
+
+err_init:
+	sdw_bus_master_delete(bus);
+	return ret;
+}
+
+static int intel_master_startup(struct sdw_intel_master_dev *master_dev)
```

Comment:
Can this be intel_master_startup(struct sdw_intel *sdw) and call it directly from sdw_intel_startup_controller()?